ci: Use ccache to cache build objects for speeding up building

bthomee opened the pull request:
## High Level Overview of Change

This change enables caching of build objects using `ccache` on Linux, macOS, and Windows.

### Context of Change

Right now, each pipeline invocation builds the source code from scratch. Although compiled Conan dependencies are cached in a remote server, the source build objects are not. We are able to further speed up our builds by leveraging `ccache`.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Performance (increase or change in throughput and/or latency)
- [ ] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [X] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

bthomee committed:
ci: Use ccache to cache build objects for speeding up building

bthomee committed:
Print ccache statistics

bthomee committed:
Update CI images

bthomee committed:
Cache the ccache dir

bthomee committed:
Use env var for the ccache dir

bthomee committed:
Update name of step

bthomee committed:
Merge branch 'develop' into bthomee/ccache

bthomee committed:
Merge branch 'develop' into bthomee/ccache

bthomee committed:
Merge branch 'develop' into bthomee/ccache

bthomee committed:
Merge branch 'develop' into bthomee/ccache

mathbunnyru commented:
> The macOS pipelines are executed on bare metal with a ccache directory located outside of the working directory

That's not true that ccache dir is outside of the working directory
ccache can and should be used on macOS (and we do it for Clio)

mathbunnyru commented:
Also, please take a look at Clio's implementation; many of the features can be implemented here as well:
- https://github.com/XRPLF/clio/blob/develop/.github/workflows/reusable-build.yml
- https://github.com/XRPLF/clio/blob/develop/cmake/Ccache.cmake (this is better to keep as a separate file, to make our approach more modular, and keep it in sync with Clio)

bthomee committed:
Merge branch 'develop' into bthomee/ccache

bthomee committed:
Tweaks

bthomee commented:
> > The macOS pipelines are executed on bare metal with a ccache directory located outside of the working directory
> 
> That's not true that ccache dir is outside of the working directory ccache can and should be used on macOS (and we do it for Clio)

Logging into our macOS GitHub runner shows that the ccache dir is the user's `Library/Caches/ccache` directory, which is outside of the working directory.

bthomee commented:
> Also, please take a look at Clio's implementation; many of the features can be implemented here as well:
> 
> * https://github.com/XRPLF/clio/blob/develop/.github/workflows/reusable-build.yml
> * https://github.com/XRPLF/clio/blob/develop/cmake/Ccache.cmake (this is better to keep as a separate file, to make our approach more modular, and keep it in sync with Clio)

I reluctantly moved the ccache CMake changes back into a separate file, simply for similarity with Clio. However, I don't see in what way this helps keep things more modular, since in my view it just increases the number of files in the directory (albeit by just 1) and it moves 5 lines of compiler-related statements into a different file than where all the compiler-related definitions are stored. 

I further looked at the ccache set up in Clio and noticed that upload+download are always enabled together, and generally not enabled for pushes into the develop and release branches or the nightly run; I simplified this by just only enabling ccache for PR commits. This also avoids me having to pass variables between multiple levels of actions.

Note that I use the `config_name` as the cache key. Since you distinguish between builds with coverage enabled, I also added it to the config name so it becomes a separate entry in the cache.

Also note that sanitizers are not yet enabled, but from the sanitizers PR I recall that they will also be part of the config name. However, @mathbunnyru, I noticed you completely disabled ccache for sanitizers - what's the motivation for doing so?

Finally, what's the motivation for not using ccache for commits into the develop or release branches? It is just to be extra safe or is there an actual known risk with ccache not producing the correct build objects?

bthomee committed:
Add include(Ccache) to CMakeLists.txt

bthomee commented:
> > > The macOS pipelines are executed on bare metal with a ccache directory located outside of the working directory
> > 
> > 
> > That's not true that ccache dir is outside of the working directory ccache can and should be used on macOS (and we do it for Clio)
> 
> Logging into our macOS GitHub runner shows that the ccache dir is the user's `Library/Caches/ccache` directory, which is outside of the working directory.

Alright, I see now that the `prepare-runner` action overrides this to be inside the working directory.

bthomee committed:
Improve build-test-config, enable ccache for MSVC

bthomee committed:
Define stats log within a step

bthomee committed:
Remove log stats as it segfaults on Debian, improve ccache command on Windows

bthomee committed:
Use multiple COMMAND statements for pipes

bthomee committed:
Remove OUTPUT_QUIET and temporarily only run on Windows

bthomee committed:
Debug

bthomee committed:
Debug

bthomee committed:
Use bash to extract ccache target on Windows

bthomee committed:
Debug

bthomee committed:
Debug

bthomee committed:
Move ccache include statements before compilers are defined

bthomee committed:
Merge branch 'develop' into bthomee/ccache

bthomee committed:
Debug

bthomee committed:
Enable all OSes again in matrix

bthomee commented:
@mathbunnyru @vlntb This PR is now ready, and ccache now works on all platforms.

mathbunnyru commented on .github/workflows/reusable-build-test-config.yml:
```diff
@@ -116,6 +117,17 @@ jobs:
 
+      - name: Restore ccache
+        if: ${{ env.CCACHE_ENABLED == 'true' }}
+        uses: actions/cache/restore@0057852bfaa89a56745cba8c7296529d2fc39830 # v4.3.0
```
You always download cache, add files to it (while building new code), and then upload a bigger cache.

This will eventually grow the cache size, and make it useless

bthomee replied:
Good point. I had hoped it would be more like Bazel/BuildBuddy where the least recently used files would be dropped automatically once a certain time/size limit was reached, but here the caches are considered as a whole unit and dropped entirely.

My original idea was to use ccache for caching across PRs, so unchanged files wouldn't need to be built again. It seems like that is not possible with the current set up. Should a separate cache be used per PR?

Per https://github.com/actions/cache/issues/1071, one comment states "This functioning is expected and cache is built on the principle that "each key uniquely identifies a cache". if a cache contents have changed and you need to update it, its key should also reflect that.". So would this mean that each commit that modifies some source files needs its own cache? That entirely defeats the purpose...

mathbunnyru replied:
> Should a separate cache be used per PR?

I don't know if that would work well in rippled's case, because it depends on how active people are pushing their PRs.
I think if there are many active PRs and they override the same cache, it's rarely going to work.

Clio doesn't create caches on PRs, only on commits on develop.
This also solves the problem with fork/non-fork.
And we add a commit hash to the cache id.

mathbunnyru commented on .github/workflows/reusable-build-test-config.yml:
```diff
@@ -116,6 +117,17 @@ jobs:
+
+      - name: Show ccache statistics before building
+        if: ${{ env.CCACHE_ENABLED == 'true' }}
+        run: ccache --show-stats ${{ runner.debug && '-vv' || '' }}
```
This will all be zeroes and only show ccache size limit, so I don't think it's needed and gives useful information

bthomee replied:
It will still show the size of the cache, even though the hits and misses are not shown (because they are zero), e.g.

```
Local storage:
  Cache size (GB): 0.43 / 5.00 ( 8.64%)
```

mathbunnyru replied:
This information is also shown after the build finishes (with new files included), and on previous upload (without them), that's why it's not that needed

mathbunnyru commented on .github/workflows/reusable-build-test-config.yml:
```diff
@@ -129,14 +141,25 @@ jobs:
+
+      - name: Save ccache
+        if: ${{ env.CCACHE_ENABLED == 'true' }}
+        uses: actions/cache/save@0057852bfaa89a56745cba8c7296529d2fc39830 # v4.3.0
```
Does this work when PR is created from fork?

bthomee replied:
Yes:
* Everyone has 10GB by default.
* Even if writing fails, the step succeeds. This is apparently a feature (although several developers have filed tickets to complain that they'd like the step to fail the pipeline if saving doesn't work).

Now, if you think it'd be better to not enable this by default in forks, I can make the change. I'm fine either way.

mathbunnyru commented on .github/workflows/reusable-build-test-config.yml:
```diff
@@ -129,14 +141,25 @@ jobs:
+        uses: actions/cache/save@0057852bfaa89a56745cba8c7296529d2fc39830 # v4.3.0
+        with:
+          path: ${{ env.CCACHE_DIR }}
+          key: ${{ inputs.config_name }}
```
What will happen if 2 PRs run simultaneously and write to the same cache key?

bthomee replied:
My understanding is that the first one to attempt the write will succeed, while the second one will fail to write with the "Unable to reserve cache with key ..., another job may be creating this cache" message.

mathbunnyru commented on cmake/Ccache.cmake:
```diff
@@ -0,0 +1,45 @@
+find_program(CCACHE_PATH "ccache")
+if (CCACHE_PATH)
```
Let's return right away if `ccache` is not found, it will make code easier to read

bthomee replied:
Good point. I applied the same to other parts of the file to further simplify it.

mathbunnyru commented on cmake/Ccache.cmake:
```diff
@@ -0,0 +1,45 @@
+find_program(CCACHE_PATH "ccache")
+if (CCACHE_PATH)
+    if (MSVC)
+        # Chocolatey uses a shim executable that we cannot use directly, in
```
Who installed ccache on Windows?
Is it part of deployment scripts?

bthomee replied:
It's part of the Terraform script that deploys the Windows runners, using `chocolatey` to install ccache. Unfortunately it cannot be installed via `pip`.

mathbunnyru replied:
Can't we do it in `prepare-runner`?
The same way we use `brew` for Mac, we'll be using `cholocatey` for Windows

bthomee replied:
Done in https://github.com/XRPLF/actions/pull/42.

mathbunnyru replied:
You have to update the usage here

bthomee replied:
Yes, I will in a follow-up PR.

I'm currently first considering taking advantage of ccache supporting Redis as a remote cache, so I can rip out the largely useless GitHub cache.

mathbunnyru commented on cmake/Ccache.cmake:
```diff
@@ -0,0 +1,45 @@
+                "CLToolExe=cl.exe"
+                "CLToolPath=${CMAKE_BINARY_DIR}"
+                "TrackFileAccess=false"
+                "UseMultiToolTask=true")
```
I have no idea what's going on here 😂

bthomee replied:
I added a link to the ccache wiki where I got most of the instructions from, which I had to further massage a bit.

mathbunnyru commented on cmake/Ccache.cmake:
```diff
@@ -0,0 +1,45 @@
+        # By default Visual Studio generators will use /Zi, which is not
+        # compatible with ccache, so tell it to use /Z7 instead. When running in
+        # CI the /Zi option will have already been stripped out, in which case
+        # the following will be a no-op, see XrplCompiler.cmake.
```
This is really good comment, thanks!
Could you please also add link(s) where you got this information from?

bthomee replied:
The main part of the comment came from the ccache wiki, see the link. I removed the last part of this comment in a recent commit as it wasn't as relevant.

bthomee committed:
Merge branch 'develop' into bthomee/ccache

bthomee committed:
Simplify Ccache.cmake

bthomee committed:
Use redis as remote cache for windows only at the moment

bthomee committed:
Fix prepare-runner path

bthomee committed:
Merge branch 'develop' into bthomee/ccache

bthomee committed:
Remove unavailable conditional

bthomee committed:
Set redis endpoint

bthomee committed:
Improve log message

bthomee committed:
Use HTTP storage backend

bthomee committed:
Use HTTP storage backend

bthomee committed:
Merge branch 'develop' into bthomee/ccache

bthomee committed:
Enable macOS and verbose cache stats

bthomee committed:
Enable linux

bthomee committed:
Update secret name

bthomee committed:
Improve comments and trigger

bthomee committed:
Use DNS name for cache endpoint

bthomee committed:
Use log file for ccache to debug issues

bthomee committed:
Fix url suffix

bthomee committed:
Move ccache log file into temp directory

bthomee committed:
Use runner.xxx instead of github.xxx

bthomee committed:
The runner.xxx is only available in steps

bthomee committed:
The runner.xxx is only available in steps

bthomee committed:
Add sloppiness to increase cache hits

bthomee committed:
Improve comments describing the ccache env vars

bthomee commented:
This PR is now ready for re-review.

Noteworthy changes and observations:
* The behavior of the GitHub cache is not what I was looking for, and I explored other storage backends.
* Ccache supports Redis as remote storage backend, which I deployed, but as it turns out AWS Elasticache requires the use of TLS, which ccache does not support. It is possible to set up a proxy that can upgrade a plain connection to a secured connection, but that would require deploying a separate EC2 instance in which case I may as well use another backend.
* Ccache also supports HTTP as remote storage backend, and suggests either using Nginx or Bazel Remote. Since I am familiar with Bazel and remote caching, and still hold out hope we can switch to Bazel some day, I deployed the latter as backend on a new EC2 instance.
* This worked fine for the Windows runner, which is deployed in the same VPC as the new EC2 instance. However, for the macOS and Linux runners this required creating a VPC peering and updating the routing table to allow these runners to connect to the cache. This has now been fully enabled.
* Since CMake is configured to move our header files into different directories (from `include/` to somewhere in the `build/` directory), I had to tell it to ignore the creation and modification date. Subsequent runs of the same workflow now result in 100% cache hits, whereas before that was 0% or a low number.

@mathbunnyru I can probably move some of the env vars to the `prepare-runner` action. In particular, if Clio would like to use the same remote backend, you are more than welcome to. I completely disabled local caching here, since there's no point when the workspace gets cleared between jobs and I'm not using the GitHub cache; this also means I don't need to clear it, since that only affects the local cache. Not having a local cache doesn't seem to meaningfully affect build times (maybe build is slightly longer, but save time by not needing to restore and save a cache anymore).

bthomee committed:
Improve ccache enabled statement

pratikmankawde commented on .github/scripts/strategy-matrix/generate.py:
```diff
@@ -232,6 +232,8 @@ def generate_strategy_matrix(all: bool, config: Config) -> list:
         )
         config_name += f"-{build_type.lower()}"
+        if "-Dcoverage=ON" in cmake_args:
+            config_name += "-coverage"
```
Is this an unrelated change?

bthomee replied:
No, this is intentional to ensure that cached objects with the coverage option enabled go into a separate namespace (which is set to the `config_name`). So if we decide to build the same distro+gcc without coverage it wouldn't potentially lead to reusing objects that were compiled with different settings.

Now, the coverage builds result in 95%+ of objects that can't be cached due to ccache not supporting the coverage compiler options, so in practice this won't be a problem. But better to be safe than sorry.

pratikmankawde commented on .github/workflows/reusable-build-test-config.yml:
```diff
@@ -79,7 +102,11 @@ jobs:
+
+      - name: Set ccache log file
+        if: ${{ env.CCACHE_ENABLED == 'true' && runner.debug }}
+        run: echo "CCACHE_LOGFILE=${{ runner.temp }}/ccache.log" >> "${GITHUB_ENV}"
```
Suggestion: Should the log file contain `CCACHE_NAMESPACE`?

bthomee replied:
The log file is produced by ccache when debug mode is enabled and includes the namespace, see e.g.

```
[2025-12-17T00:40:33.425210 39262] Config: (environment) namespace = rhel-10-clang-any-amd64-release
```

A full debug run can be found here: https://github.com/XRPLF/rippled/actions/runs/20286243246/job/58264275051?pr=6104. Expand the "Show ccache statistics" to see what the log file contains.

pratikmankawde replied:
Sorry, I mistyped.
I wanted to ask: _Should the log file **name** contain `CCACHE_NAMESPACE`?_ (to avoid overwriting)

bthomee replied:
Ah, I see what you mean. No, that shouldn't matter as the log file only exists for the duration of the job.

The runner temp directory is automatically cleared out after the job finishes. Even if we would put it somewhere in the workspace, the same would happen. We could upload it as an artifact, but since its contents are printed to the output when debug mode is enabled, I think that's good enough.

Now, I'm open to enabling - in a follow-up PR - the `CCACHE_DEBUG` option, which produces an entire tree of files that allow you to actually dig into differences between executions so you can figure out why something resulted in a cache miss. Right now the log file isn't sufficient to actually do that. For instance, the RHEL 10 runs take 10 mins to build and result in almost no cache hits, so there's something non-deterministic there that isn't affecting other distro/compiler combinations. But I propose to get this merged first so we can take advantage of the speed-up on most configurations, and then optimize it later.

pratikmankawde reviewed (approved):
LGTM

bthomee committed:
Merge branch 'develop' into bthomee/ccache

bthomee committed:
Move ccache_enabled to the upstream repos, simplify logic

bthomee committed:
Use boolean input directly

mathbunnyru commented on .github/workflows/reusable-build-test-config.yml:
```diff
@@ -79,7 +101,11 @@ jobs:
+          disable_ccache: ${{ !inputs.ccache_enabled }}
+
+      - name: Set ccache log file
+        if: ${{ inputs.ccache_enabled && runner.debug }}
```
```suggestion
        if: ${{ inputs.ccache_enabled && runner.debug == '1' }}
```

Let's bee consistent with what's written next

bthomee replied:
Interesting, I thought it was a boolean variable, but the documentation indeed shows it is a string that always has the value '1' when enabled.

mathbunnyru commented on .github/workflows/on-trigger.yml:
```diff
@@ -66,6 +66,12 @@ jobs:
       matrix:
+        # Enable ccache only for events targeting the XRPLF repository, since
+        # other accounts will not have access to our remote cache storage.
+        # However, we do not enable ccache for events targeting the master or a
+        # release branch, to protect against the rare case that the output
+        # produced by ccache is not identical to a regular compilation.
```
There is another, and in my opinion more important, reason to disable ccache for any important executions in this implementation.

Correct me if I'm wrong, but anyone who is able to run something in our VPC can have write access to the cache.
This includes all `heavy` runners, in all repos/forks and includes former employees, who once set runners, and now have them.

Attack: poison redis ccache, and replace the correct object file with the one which has malicious code.
And you can even do it in a private fork, and it will be difficult to trace where the attack came from (and will probably be too late).

I don't know how PRs are set-up for external contributors, but if they are run automatically, then attacker will be able to poison cache from PR, without setting up runners at all.

Note: actions/cache doesn't have this issue, because it only has write access when the GITHUB_TOKEN has the write access.

Important implication would be: binaries built/tested in PRs can't be trusted at all.
And if someone deploys from a branch, this can't also be trusted.

What do you think about this?

> I understand this problem is not exactly about these lines, and more about approach in general, but GitHub doesn't allow to reply to general comments in a thread-like fashion, so I left it as comment on these lines.

bthomee replied:
Yes, the risk of cache poisoning is why I disable the cache completely for all release-related commits. Indeed, a ramification of this is that no one should ever deploy from a non-release branch, as you pointed out.

The way the deploy pipelines are set up (currently in GitLab based on a mirror, but they will soon be moved to this repo) is that release packages are only built from the release branch. So if someone wants to do the wrong thing - build from a non-release branch - then they have to make more effort to manually create and then deploy these packages. The risk & ability to do so, however, will not meaningfully increase as a result of this change, since it requires an insider who already has the necessary permissions to do damage right now. Note that as part of moving the packaging pipeline over to here I'm planning to leverage "environments" that require an additional approval before a package is released and/or deployed - giving another chance to sanity check.

As for external bad actors who intentionally try to poison the cache via a PR, the pipelines are configured to require a maintainer to approve them to run. A maintainer should never let them run without checking the code first. If a maintainer lets these pipelines run, I suspect that the risk of cache poisoning might be similar when using the GitHub cache, since the GITHUB_TOKEN presumably would get the maintainer's permissions. There's a risk of approved contributors with write access whose pipelines will automatically run for PRs, but as long as releases are not built with the cache enabled we should be fine.

mathbunnyru replied:
> If a maintainer lets these pipelines run, I suspect that the risk of cache poisoning might be similar when using the GitHub cache, since the GITHUB_TOKEN presumably would get the maintainer's permissions.

I don't think it works this way.
There is a restriction that GitHub implements when running from PR: https://docs.github.com/en/actions/reference/workflows-and-actions/dependency-caching#restrictions-for-accessing-a-cache
So, the PR will only be able to write to the cache scope of the PR, so no one else outside of this PR can be affected.

While with this solution, there is no such restriction

bthomee replied:
Ack, I recognize the differences. The question is whether the risk is acceptable. I confirmed with our platform team that critical nodes (UNL validator, hubs) run the binary from packages produced from the release branch. The XRPL  is designed to handle malicious nodes, so even if a regular node is updated with a tampered package it won't affect the stability of the network. I'll discuss it further with them and you offline to see if there are other risk vectors.

bthomee committed:
Update .github/workflows/reusable-build-test-config.yml

Co-authored-by: Ayaz Salikhov <mathbunnyru@users.noreply.github.com>

bthomee committed:
Merge branch 'develop' into bthomee/ccache

mathbunnyru reviewed (changes requested):
For some reason, `ccache` didn't work for `ubuntu-jammy-gcc-12-arm64-debug`: https://github.com/XRPLF/rippled/actions/runs/20467375572/job/59486838790?pr=6104

Could you please check why?

bthomee committed:
Merge branch 'develop' into bthomee/ccache

bthomee committed:
Print ccache version

bthomee committed:
Update CI image shas

bthomee commented:
@mathbunnyru the updated ccache version has resulted in both the coverage and ubuntu-jammy builds successfully using the cache 😄

mathbunnyru commented on .github/actions/print-env/action.yml:
```diff
@@ -11,6 +11,9 @@ runs:
         set
 
+        echo 'Checking ccache version.'
+        ccache --version
```
ccache, cmake and conan should be printed for all OSs, let's remove duplication here and have 3 steps:
- common
- Windows
- Linux and macOS

bthomee replied:
Done. I added the common ("all") as a last step. Windows, Linux, and macOS start with printing out env vars, and then continue with tool versions; the common step also prints out versions.

mathbunnyru commented on .github/workflows/reusable-build-test-config.yml:
```diff
@@ -10,8 +10,14 @@ on:
+        description: "Whether to enable ccache."
+        required: false
+        type: boolean
+        default: false
```
Let's not create an input variable here; it can be determined without having to pass it, you can have an env variable to not have duplicates

mathbunnyru commented on .github/workflows/reusable-build-test-config.yml:
```diff
@@ -21,8 +27,8 @@ on:
         description: "The CMake target to build."
-        type: string
         required: true
+        type: string
```
I highly suggest in the future to make refactoring changes in a separate PR, especially when it touches completely unrelated code.

This:
- allows to fix it in all the places, and make it consistent everywhere
- keeps current PR more focused
- makes it easier to review, and the refactoring change can be merged much faster

bthomee committed:
Check configuration steps

mathbunnyru commented:
Could you please check why [`rhel-10-clang-any-amd64-release`](https://github.com/XRPLF/rippled/pull/6104#logs) still takes more than 8 minutes?

bthomee commented:
> Could you please check why [`rhel-10-clang-any-amd64-release`](https://github.com/XRPLF/rippled/pull/6104#logs) still takes more than 8 minutes?

I have no obvious explanation yet but I'll keep digging. If you're ok with it then we can go ahead and merge this PR, and I'll figure it out afterwards. I'm also suspecting there may be additional configs that have the same issue and we won't discover until the nightly run kicks those off.

mathbunnyru commented:
> > Could you please check why [`rhel-10-clang-any-amd64-release`](https://github.com/XRPLF/rippled/pull/6104#logs) still takes more than 8 minutes?
> 
> I have no obvious explanation yet but I'll keep digging. If you're ok with it then we can go ahead and merge this PR, and I'll figure it out afterwards. I'm also suspecting there may be additional configs that have the same issue and we won't discover until the nightly run kicks those off.

Let's update images to the one with ccache built from source (it might help, but might not).
Even if it doesn't, feel free to merge this (and figure it out afterwards).

mathbunnyru reviewed (changes requested):
Ah, I think this solution brought a few problems back:
- Who is responsible for cleaning the cache? If no one does it, won't it grow without limit, exhausting the disk space?
- Stats won't show what matters, because they will always add up. If you build let's say 150 files, you want to see how many of these files hit/missed the cache. Instead, I think it would show how many of all the runs in all commits and PRs this configuration so far hit/missed the cache (and this information is not useful)

I'm sorry to bring this up late; I've just thought about it.

bthomee committed:
Update CI image shas

bthomee commented:
> Ah, I think this solution brought a few problems back:
> 
> * Who is responsible for cleaning the cache? If no one does it, won't it grow without limit, exhausting the disk space?
> * Stats won't show what matters, because they will always add up. If you build let's say 150 files, you want to see how many of these files hit/missed the cache. Instead, I think it would show how many of all the runs in all commits and PRs this configuration so far hit/missed the cache (and this information is not useful)
> 
> I'm sorry to bring this up late; I've just thought about it.

Ccache is configured to limit the cache size of each namespace to 5GB (the default), and I configured the remote HTTP cache (based on Bazel Remote) to have a maximum size of slightly less than the disk size I provided it (~1TB). Together they will limit the amount of data that can be cached, with LRU data being removed. In practice I don't think we'll hit the remote cache disk size, because we are running under 200 configurations.

We can chat further about what stats are useful to show, and if we resetting it first like you've done with the Clio pipelines gives us more valuable info. However, let's get this merged first so we can start taking advantage of faster build times.

mathbunnyru commented:
> > Ah, I think this solution brought a few problems back:
> > 
> > * Who is responsible for cleaning the cache? If no one does it, won't it grow without limit, exhausting the disk space?
> > * Stats won't show what matters, because they will always add up. If you build let's say 150 files, you want to see how many of these files hit/missed the cache. Instead, I think it would show how many of all the runs in all commits and PRs this configuration so far hit/missed the cache (and this information is not useful)
> > 
> > I'm sorry to bring this up late; I've just thought about it.
> 
> Ccache is configured to limit the cache size of each namespace to 5GB (the default), and I configured the remote HTTP cache (based on Bazel Remote) to have a maximum size of slightly less than the disk size I provided it (~1TB). Together they will limit the amount of data that can be cached, with LRU data being removed. In practice I don't think we'll hit the remote cache disk size, because we are running under 200 configurations.

Nice, thanks 👍 

> We can chat further about what stats are useful to show, and if we resetting it first like you've done with the Clio pipelines gives us more valuable info. However, let's get this merged first so we can start taking advantage of faster build times.

Ok, let's go ahead and merge this
